kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java

```diff
@@ -81,32 +114,50 @@ public List<Path> getCorrespondingFiles(Path path) {
         if (this == CheckpointInstance.MAX_VALUE) {
             throw new IllegalStateException("Can't get files for CheckpointVersion.MaxValue.");
         }
+
+        // This is safe because the only way to construct a V2 CheckpointInstance is with the path.
+        if (format == CheckpointFormat.V2) {
+            return Collections.singletonList(filePath.get());
+        }
+
         return numParts
-            .map(parts -> FileNames.checkpointFileWithParts(path, version, parts))
-            .orElseGet(() ->
-                Collections.singletonList(FileNames.checkpointFileSingular(path, version)));
+                .map(parts -> FileNames.checkpointFileWithParts(path, version, parts))
+                .orElseGet(() ->
+                        Collections.singletonList(
+                                FileNames.checkpointFileSingular(path, version)));
     }
 
     /**
      * Comparison rules:
      * 1. A CheckpointInstance with higher version is greater than the one with lower version.
-     * 2. For CheckpointInstances with same version, a Multi-part checkpoint is greater than a
+     * 2. A CheckpointInstance for a V2 checkpoint is greater than a classic checkpoint (to filter
+     *    avoid selecting the compatibility file) or a multipart checkpoint.
+     * 3. For CheckpointInstances with same version, a Multi-part checkpoint is greater than a
      *    Single part checkpoint.
-     * 3. For Multi-part CheckpointInstance corresponding to same version, the one with more
+     * 4. For Multi-part CheckpointInstance corresponding to same version, the one with more
      *    parts is greater than the one with less parts.
+     * 5. For V2 checkpoints, use the file path to break ties.
      */
     @Override
     public int compareTo(CheckpointInstance that) {
         if (version == that.version) {
-            return Long.compare(numParts.orElse(1), that.numParts.orElse(1));
+            if (format == that.format) {
+                if (numParts.orElse(1).equals(that.numParts.orElse(1)) && format ==
+                        CheckpointFormat.V2 && filePath.isPresent() && that.filePath.isPresent()) {
+                    return filePath.get().getName().compareTo(that.filePath.get().getName());
+                }
```

Comment:
can we just have a swtich on `format` and comparison for each one separately at this point on the format? The If condition is complicated to understand.